### DISCUSSION TITLE
How can I get the pidSV value and send the value out through websockets ?

### DISCUSSION THREAD

**Discussion by sakunamary** [artisan-roaster-scope/artisan#866] [Q&A] [answered]
Hi ALL guys:
     I get a small roaster called Sniper M1 which made by Kaleido with PAD(you can some info from here .http://www.kaleido-coffee.com/product/hot-sale-giesen-w1a-hot-top-kuban-coffee-roaster-price). And I am trying to control the little roaster via ESP8266 by websockets  with Artisan . The control pad communicate with the drummer by serial protocol commands,and the little roaster has it own internal PID controller ,you just need to send the Target temperature  ,the internal PID will control the heater power automatically .you can get the ET BT fans value from the serial data easily. 
after the coding works and testing works
     1.ESP8266 WIFI config is done.
     2.ESP8266 get the data from drummer is done .
     3.pad send and receive data from frummer is done .
     4.ESP8266 send control command is done .
    BUT......
    5.ESP8266 get the Artisan pidSV is NOT done ...

here is my artisan-settings :

[M1artisan-settings_websockets.zip](https://github.com/artisan-roaster-scope/artisan/files/8456456/M1artisan-settings_websockets.zip)

I try to set Events 'TargetC'  to control the drummer heating target temperature .
![微信截图_20220409160634](https://user-images.githubusercontent.com/3532903/162562811-054ef16b-4c4b-4ac3-9d17-74305ddb8633.png)

And I also set two button "TargetC web" and "TargetC pidsv" to send  out commands ,but not work out .I mean send the command function is OK ,but the button only can send the fixed value ,Target temperature will be adjust in the roast process  .

![微信截图_20220409161014](https://user-images.githubusercontent.com/3532903/162562954-668e5a7f-d69a-49f3-b46b-3f7e9ed6876e.png)

So I try to setup a slider bar . the slider bar send command separately (pidSV and websocket -send()) the value is OK ,the Drummer react correctly .But I can not send the value in the same time ,so that the pidSV can not sync with the websockets command .

![微信截图_20220409161557](https://user-images.githubusercontent.com/3532903/162563095-630c6394-07a5-4bcb-aaa2-c8812df7e677.png)

HOW CAN I pass through the pidSV value to the websocket -send( ) command ?  
Is it possible to get the pidSV value from any function or modules ?
OR  add the multi-commands function in slider bar just like button?

Thanks!

**Comment by MAKOMO** [comment]
I don't understand your project at all, but note that those Command strings accept sequences of commands separated by semicolon. Thus you can use something like this

`send({{TargetC:{}}});send({{BurnerValue:{},AnyOtherLabel:12}})`

**Comment by sakunamary** [comment]
Thanks for your answer!!!

My （Kaleido M1） roaster control board can hold the target temperature by itself（Internal PID controller）, all I need to do is send  a target temperature through websocket command.

For your understanding , just like Arduino TC4 ,In the PID background mode Artisan working like this :

<img width="567" alt="WX20220411-083607@2x" src="https://user-images.githubusercontent.com/3532903/162647238-68a8a68c-c27d-4fa6-b878-7d4171e1bc37.png">

 1.Artisan load a background curves ,and  then Artisan send out Serial command  "PID,SV,90"  , the target BT 90 degree
  will be send to   TC4 ,
 2.TC4 get the SV value , the  controller will do the rest thing to lock the BT at SV .
     and go to  next loop ,
 Artisan  read the BT data on the background curves  again ,and send another new  PID SV value , like PID,SV,100(for example ) to TC4 .
.......
  After many loops, as the result , the BT curves will look exactly  like the background curves .

In my case my roaster what it doing is same as TC4 receive command "PID,SV,xxx", The roaster need to know the PID SV ,which is the most important  parms,Artisan can send "PID,SV,xxx" to TC4 .

workflow like this:
 1.Artisan read background curves BT value --> 2.send BT value through websocket command  send(TargetC value  ) --> 3.roaster get the Target C and lock and hold the BT on the roaster.

 But I don't how to  send a "PID,SV,xxx" in websocket command , or websocket can doing like or not ?




 p.s 
  I am not understanding the meaning of following  command:
    send({{BurnerValue:{},AnyOtherLabel:12}})
   what dose "AnyOtherLabel:12" mean ?  12 mean "button 12" ?  

MANY THANKS again!!!
Best wishes!!!

**Reply by MAKOMO** [reply]
Artisan is currently not supporting external PIDs connected via WebSockets. External PID support is implemented only for TC4, MODBUS and S7 connections. Sorry.

**Reply by poundy** [reply]
And that means you should be considering implementing Modbus over TCP/IP on your ESP node.

**Reply by sakunamary** [reply]
> And that means you should be considering implementing Modbus over TCP/IP on your ESP node.

Modbus is too difficult to me  ,I have no idea about Modbus .... LOL....

Thanks for your sharing!!

**Reply by sakunamary** [reply]
> Artisan is currently not supporting external PIDs connected via WebSockets. External PID support is implemented only for TC4, MODBUS and S7 connections. Sorry.

Oh .got it , I should try another path, making  an interpreter for TC4 to M1 which base on TC4 protocol ,so that Artisan can send PID SV to the roaster.
anyway ,thank you  for your answer !

**Reply by MAKOMO** [reply]
Sorry, my answer was incomplete (or wrong). Artisan has the possibility to send out a custom command at regular time intervals or at every sampling step. This is configured in the Events tab (`Config >> Events`, 1st tab). If you choose there to send out a WebSocket with the following specification

```
send({"target": {BTB}})
```

Artisan will send out the background profiles current BT reading (bound to the variable `{BTB}`).

<img width="1002" alt="Screenshot 2022-04-29 at 07 56 41" src="https://user-images.githubusercontent.com/13909011/165891908-c700f0ac-b399-4ca9-87d1-71092bc98c36.png">

With this Artisan will send out something like this to the connected WebSocket:

```
{'target': 190.81851178294332, 'id': 9140, 'roasterID': 0}
```

In case Artisan is not recording or there is no background profile data available it will send out `-1`.

Sorry that this mechanism just came now to my mind. Artisan has many features and functions!

**Reply by sakunamary** [reply]
Thanks Bros！
What  a GOOOOOOD news !But  I will try it later, because I'm  still waiting the PCBs of my project shipping to me . LOL
 
Best wishes!

**Reply by MAKOMO** [reply]
Note that this is not as "comfortable" as using MODBUS/S7 in an external PID setting as the PID dialog is bypassed here. Thus you cannot switch PID mods from manual/ramp-soak/background from within the Artisan PID dialog as in the MODBUS/S7 case.

**Comment by sakunamary** [comment, marked as answer]
Just Mark down the Answer:

Artisan has the possibility to send out a custom command at regular time intervals or at every sampling step. This is configured in the Events tab (`Config >> Events`, 1st tab). If you choose there to send out a WebSocket with the following specification
 ```
send({"target": {BTB}})
 ```

Artisan will send out the background profiles current BT reading (bound to the variable `{BTB}`).
<img alt="Screenshot 2022-04-29 at 07 56 41" width="1002" src="https://user-images.githubusercontent.com/13909011/165891908-c700f0ac-b399-4ca9-87d1-71092bc98c36.png">
With this Artisan will send out something like this to the connected WebSocket:
```
{'target': 190.81851178294332, 'id': 9140, 'roasterID': 0}
```
   
In case Artisan is not recording or there is no background profile data available it will send out `-1`.

After coding and Testing , Now is working , Look at following Pic .
![IMG_9844](https://user-images.githubusercontent.com/3532903/166862538-58a00001-8a52-4ab1-b74a-d5606d7d657f.jpeg)
